File: uk101-replica/Generate/old/process.lua
---
---			6502 Processor 
---

local modes = { }

modes.a = { eac = "FETCHWORD();eac = temp16;", cycles = 0, desc = "@2" }
modes.ax = { eac = "FETCHWORD();eac = temp16+x;", cycles = 0, desc = "@2,X" }
modes.ay = { eac = "FETCHWORD();eac = temp16+y;", cycles = 0, desc = "@2,Y" }
modes.z = { eac = "eac = FETCH();", cycles = -1, desc = "@1" }
modes.zx = { eac = "eac = (FETCH()+x) & 0xFF;", cycles = 0, desc = "@1,X" }
modes.zy = { eac = "eac = (FETCH()+y) & 0xFF;", cycles = 0, desc = "@1,Y" }
modes.i = { eac = "", cycles = -2, desc = "#@1" }
modes.id = { eac = "temp16 = FETCHWORD();eac = READWORD(temp16);", cycles = 2, desc = "(@2)" }
modes.ix = { eac = "temp8 = FETCH()+x;eac = READWORD01(temp8);", cycles = 3, desc = "(@1,X)" }
modes.iy = { eac = "temp8 = FETCH();eac = READWORD01(temp8)+y;", cycles = 2, desc = "(@1),Y" }

local codeBlock = {} 			
local mnemonicList = {}
local codeList = {}

function defineOpcode(opcode,mnemonic,code,cycles) 
	assert(mnemonicList[opcode] == nil,"Duplicate")
	mnemonicList[opcode] = mnemonic
	codeList[opcode] = "CYCLES("..cycles..");"..code
	--print(("%02x"):format(opcode),mnemonic,code)
end

local src = io.open("6502.def")
for r in src:lines() do
	local n = r:find("//") if n ~= nil then r = r:sub(1,n-1) end
	r = r:match("^%s*(.*)$")
	while r:match("%s$") do r = r:sub(1,#r-1) end
	if r:sub(1,1) == ":" then
		codeBlock[#codeBlock+1] = r:sub(2)
		r = ""
	end
	if #r ~= 0 then													-- simple address mode instructions
		local mnemonicBase,cycles,opcodes,code = r:match('^"([%w%s%@]+)"%s+(%d+)%s+([%(%)%w%,%:]+)%s+(.*)$')
		assert(mnemonicBase ~= nil,"Syntax error .. " .. r)
		if #opcodes == 2 then 
			defineOpcode(tonumber(opcodes,16),mnemonicBase,code,cycles)
		else
			opcodes = opcodes:match("^%((.*)%)$") .. ","			-- multiple address mode instructions
			assert(opcodes ~= nil,"Syntax error in opcodes .. " .. r)
			repeat 
				local mode,op
				mode,op,opcodes = opcodes:match("^(%w+)%:(%w+)%,(.*)$")
				local descr = modes[mode:lower()]
				local body = code
				if mode:lower():sub(1,1) == "z" then				-- optimise Z,ZX,ZY to use READ01()
					body = body:gsub("READ%(","READ01("):gsub("WRITE%(","WRITE01(")
				end
				if mode:lower() == "i" then 						-- if immediate no EAC,but use FETCH() to retrieve operand.
					body = body:gsub("READ%(eac%)","FETCH()")
				end
				op = tonumber(op,16)
				defineOpcode(op,mnemonicBase:gsub("@O",descr.desc),body:gsub("@EAC",descr.eac),cycles+descr.cycles)
			 until #opcodes == 0 
		end
	end
end

for i = 0,255 do
	if mnemonicList[i] == nil then mnemonicList[i] = ("byte %02X"):format(i) end
	mnemonicList[i] = mnemonicList[i]:lower()
end

hCode = io.open("6502_autocode.h","w")
for _,l in ipairs(codeBlock) do hCode:write(l.."\n") end
hCode:close()

hMnemonics = io.open("6502_mnemonics.h","w")
mnelist = '"'..table.concat(mnemonicList,'","',0,255)..'"'
hMnemonics:write("static char *_mnemonics[256] = { "..mnelist.."};")
hMnemonics:close()

hCase = io.open("6502_opcodes.h","w")
for i = 0,255 do
	if codeList[i] ~= nil then
		hCase:write(("case 0x%02x: /* $%02x:%s */\n"):format(i,i,mnemonicList[i]))
		hCase:write("    " .. codeList[i]..";break;\n")
	end
end
hCase:close()

File: superboard-emulator/processor/process.py
import re

modes = {}

modes["a"] = { "eac":"FETCHWORD();eac = temp16;", "cycles":0, "desc":"@2" }
modes["ax"] = { "eac":"FETCHWORD();eac = temp16+x;", "cycles":0, "desc":"@2,X" }
modes["ay"] = { "eac":"FETCHWORD();eac = temp16+y;", "cycles":0, "desc":"@2,Y" }
modes["z"] = { "eac":"eac = FETCH();", "cycles":-1, "desc":"@1" }
modes["zx"] = { "eac":"eac = (FETCH()+x) & 0xFF;", "cycles":0, "desc":"@1,X" }
modes["zy"] = { "eac":"eac = (FETCH()+y) & 0xFF;", "cycles":0, "desc":"@1,Y" }
modes["i"] = { "eac":"", "cycles":-2, "desc":"#@1" }
modes["id"] = { "eac":"FETCHWORD();eac = READWORD(temp16);", "cycles":2, "desc":"(@2)" }
modes["ix"] = { "eac":"temp8 = FETCH()+x;eac = READWORD01(temp8);", "cycles":3, "desc":"(@1,X)" }
modes["iy"] = { "eac":"temp8 = FETCH();eac = READWORD01(temp8)+y;", "cycles":2, "desc":"(@1),Y" }

mnemonics = [ None ] * 256
codeList =  [ None ] * 256

def defineOpcode(opcode,mnemonic,code,cycles):
	assert mnemonics[opcode] is None
	mnemonics[opcode] = mnemonic.lower() 
	codeList[opcode] = "CYCLES({0});{1}".format(cycles,code)
	#print("{1:02x} {0} {2}".format(mnemonics[opcode],opcode,codeList[opcode]))

src = open("6502.def").readlines()
src = [x if x.find("//") < 0 else x[:x.find("//")] for x in src]
src = [x.strip().replace("\t"," ") for x in src if x.strip() != ""]

open("__6502support.h","w").write("\n".join([x[1:] for x in src if x[0] == ':']))

src = [x for x in src if x[0] != ':']

for l in src:
	print(l)
	m = re.match("^\"([\w\s\%@]+)\"\s+(\d+)\s+([\(\)\w\:\,]+)\s+(.+)",l)
	assert m is not None

	if len(m.group(3)) == 2:
		defineOpcode(int(m.group(3),16),m.group(1),m.group(4),int(m.group(2)))
	else:
		opcodes = re.match("^\((.*)\)$",m.group(3))
		for mode in opcodes.group(1).split(","):
			m2 = re.match("^(\w+)\:([0-9A-F]+)$",mode)
			assert(m2 is not None)
			mode = m2.group(1).lower()
			body = m.group(4)
			assert mode in modes
			if mode[0] == 'z':
				body = body.replace("READ(","READ01(").replace("WRITE(","WRITE01(")
			if mode == 'i':
				body = body.replace("READ(eac)","FETCH()")
			mnemonic = m.group(1).replace("@O",modes[mode]["desc"])
			body = body.replace("@EAC",modes[mode]["eac"],int(m.group(2)))
			defineOpcode(int(m2.group(2),16),mnemonic,body,int(m.group(2))+modes[mode]["cycles"])

for i in range(0,256):
	if mnemonics[i] is None:
		mnemonics[i] = "byte {0:02x}".format(i)

m = ['"'+m+'"' for m in mnemonics]
open("__6502mnemonics.h","w").write("static const char *_mnemonics[] = { "+",".join(m) + "};")

handle = open("__6502opcodes.h","w")
for i in range(0,256):
	if codeList[i] is not None:
		handle.write("case 0x{0:02x}: /* ${0:02x} {1} */\n".format(i,mnemonics[i]))
		handle.write("\t{0};break;\n".format(codeList[i]).replace(";;",";"))

